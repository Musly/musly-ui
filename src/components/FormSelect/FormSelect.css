.select {
  display: flex;

  &:hover {
    .select__label {
      @apply text-gray-600;
    }

    .select__label--secondary {
      @apply text-gray-400;
    }

    .select__label--error {
      @apply text-red-500;
    }

    .select__select {
      @apply border-blue-300;
    }

    .select__select--secondary {
      @apply border-yellow-300;
    }

    .select__select--error {
      @apply border-red-500;
    }
  }
}

.select--horizontal {
  flex-direction: row;
  align-items: center;
}

.select--vertical {
  flex-direction: column;
}

.select__label {
  font-weight: 600;
  cursor: pointer;
  transition-property: color;
  transition-duration: 300ms;

  @apply mb-1 ease-sharp text-gray-500 text-sm uppercase;
}

.select__label--horizontal {
  width: 20rem;
}

.select__label--secondary {
  @apply text-gray-300;
}

.select__label--error {
  @apply text-red-500;
}

.select__select {
  cursor: pointer;
  width: 100%;

  @apply form-select focus:outline-none focus:ring-2 border
    ring-opacity-20 bg-white py-3 px-4 rounded-md transition duration-200 ease-sharp
    ring-blue-500 border-gray-300 focus:border-blue-500;

  &::placeholder {
    @apply text-gray-300;
  }
}

.select__select--secondary {
  @apply ring-yellow-500 border-gray-100 focus:border-yellow-500;
}

.select__select--error {
  @apply ring-red-500 border-red-500 focus:border-red-500;
}

.select__error {
  @apply text-red-500 mt-1 text-sm;
}
