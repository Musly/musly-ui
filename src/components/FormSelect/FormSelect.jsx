/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import './FormSelect.css';

export function FormSelect (props) {
  const {
    horizontal, error, secondary, label, value, onChange, placeholder, options, ...rest
  } = props;

  const wrapperClasses = classNames({
    select: true,
    'select--horizontal': horizontal,
    'select--vertical': !horizontal,
  });

  const labelClasses = classNames({
    select__label: true,
    'select__label--error': !!error,
    'select__label--secondary': !error && secondary,
    'select__label--horizontal': horizontal,
  });

  const inputClasses = classNames({
    select__select: true,
    'select__select--error': !!error,
    'select__select--secondary': !error && secondary,
  });

  return (
    <label className={wrapperClasses}>
      {label && (
        <span className={labelClasses}>
          {label}
        </span>
      )}
      <select onChange={onChange} value={value} {...rest} className={inputClasses}>
        {options.map((option) => (
          <option
            value={option.value}
            key={option.value}
            selected={value === option.value}
          >
            {option.label}
          </option>
        ))}
      </select>
    </label>
  );
}

FormSelect.propTypes = {
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    label: PropTypes.string,
  })).isRequired,
  onChange: PropTypes.func,
  label: PropTypes.string,
  secondary: PropTypes.bool,
  horizontal: PropTypes.bool,
  error: PropTypes.string,
  placeholder: PropTypes.string,
};

FormSelect.defaultProps = {
  onChange: null,
  label: null,
  secondary: false,
  horizontal: false,
  error: null,
  placeholder: null,
};
