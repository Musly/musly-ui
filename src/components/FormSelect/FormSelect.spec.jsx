import React from 'react';
import { shallow } from 'enzyme';
import { FormSelect } from './FormSelect';

const options = [
  { value: 1, label: 'First' },
  { value: 2, label: 'Second' },
  { value: 3, label: 'Third' },
];

describe('<FormSelect />', () => {
  it('it should render without label', () => {
    const onChange = jest.fn();
    const result = shallow(<FormSelect value={1} options={options} onChange={onChange} />);
    expect(result.find('.select__select')).toExist();
    expect(result.find('.select__label')).not.toExist();
    expect(result.find('option').get(0).props.selected).toBe(true);
    expect(result.find('option').get(1).props.selected).toBe(false);
    expect(result.find('option').get(2).props.selected).toBe(false);
    result.find('select').simulate('change');
    expect(onChange).toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });

  it('should render with label', () => {
    const result = shallow(<FormSelect label="Test label" value={2} options={options} />);
    expect(result.find('.select__select')).toExist();
    expect(result.find('.select__label')).toExist();
    expect(result.find('option').get(0).props.selected).toBe(false);
    expect(result.find('option').get(1).props.selected).toBe(true);
    expect(result.find('option').get(2).props.selected).toBe(false);
    expect(result).toMatchSnapshot();
  });
});
