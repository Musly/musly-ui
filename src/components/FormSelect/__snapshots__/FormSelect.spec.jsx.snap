// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSelect /> it should render without label 1`] = `
<label
  className="select select--vertical"
>
  <select
    className="select__select"
    onChange={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    value={1}
  >
    <option
      key="1"
      selected={true}
      value={1}
    >
      First
    </option>
    <option
      key="2"
      selected={false}
      value={2}
    >
      Second
    </option>
    <option
      key="3"
      selected={false}
      value={3}
    >
      Third
    </option>
  </select>
</label>
`;

exports[`<FormSelect /> should render with label 1`] = `
<label
  className="select select--vertical"
>
  <span
    className="select__label"
  >
    Test label
  </span>
  <select
    className="select__select"
    onChange={null}
    value={2}
  >
    <option
      key="1"
      selected={false}
      value={1}
    >
      First
    </option>
    <option
      key="2"
      selected={true}
      value={2}
    >
      Second
    </option>
    <option
      key="3"
      selected={false}
      value={3}
    >
      Third
    </option>
  </select>
</label>
`;
